src/components/shapes/draw.js

```diff
@@ -579,6 +599,208 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
     }
 }
 
+function drawLabel(gd, index, options, shapeGroup) {
+    // Remove existing label
+    shapeGroup.selectAll('.shape-label').remove();
+
+    // If no label, return
+    if(!options.label) return;
+
+    var labelGroupAttrs = {
+        'data-index': index,
+    };
+    var text = options.label.text;
+    var font = options.label.font;
+
+    var labelTextAttrs = {
+        'data-notex': 1
+    };
+
+    var labelGroup = shapeGroup.append('g')
+        .attr(labelGroupAttrs)
+        .classed('shape-label', true);
+    var labelText = labelGroup.append('text')
+        .attr(labelTextAttrs)
+        .classed('shape-label-text', true)
+        .text(text);
+
+    // If x0, x1, y0, y1 are defined explicitly, use those values
+    // Otherwise, use shape bounding box
+    var shapex0, shapex1, shapey0, shapey1;
+    if('x0' in options && 'x1' in options && 'y0' in options && 'y1' in options) {
+        // Setup conversion functions
+        var xa = Axes.getFromId(gd, options.xref);
+        var xRefType = Axes.getRefType(options.xref);
+        var ya = Axes.getFromId(gd, options.yref);
+        var yRefType = Axes.getRefType(options.yref);
+        var x2p = helpers.getDataToPixel(gd, xa, false, xRefType);
+        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);
+
+        shapex0 = x2p(options.x0);
+        shapex1 = x2p(options.x1);
+        shapey0 = y2p(options.y0);
+        shapey1 = y2p(options.y1);
+    } else {
+        // Get shape bounding box
+        var shapeBBox = shapeGroup.selectAll('path').node().getBoundingClientRect();
+        shapex0 = shapeBBox.x;
+        shapex1 = shapeBBox.right;
+        shapey0 = shapeBBox.y;
+        shapey1 = shapeBBox.bottom;
+    }
+
+    // Handle `auto` angle
+    var textangle = options.label.textangle;
+    if(textangle === 'auto') {
+        if(options.type === 'line') {
+            // Auto angle for line is same angle as line
+            textangle = calcTextAngle(shapex0, shapey0, shapex1, shapey1);
+        } else {
+            // Auto angle for all other shapes is 0
+            // The default angle for other shapes is 0 anyway, so there's no reason to
+            // explicitly set the angle to `auto`, this is just covering an edge case
+            textangle = 0;
+        }
+    }
+
+    // Do a fake render so we can get the text bounding box height
```

Comment:
@hannahker wondering if you could also use a similar technique in #6428 to handle multi-line text and improve the positioning?